sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java

```diff
@@ -0,0 +1,176 @@
+  public interface BigQueryClusteringITOptions
+      extends TestPipelineOptions, ExperimentalOptions, BigQueryOptions {
+    @Description("Table to read from, specified as " + "<project_id>:<dataset_id>.<table_id>")
+    @Default.String(WEATHER_SAMPLES_TABLE)
+    String getBqcInput();
+
+    void setBqcInput(String value);
+  }
+
+  static class KeepStationNumberAndConvertDate extends DoFn<TableRow, TableRow> {
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      String day = (String) c.element().get("day");
+      String month = (String) c.element().get("month");
+      String year = (String) c.element().get("year");
+
+      TableRow row = new TableRow();
+      row.set("station_number", c.element().get("station_number"));
+      row.set("date", String.format("%s-%s-%s", year, month, day));
+      c.output(row);
+    }
+  }
+
+  static class ClusteredDestinations extends DynamicDestinations<TableRow, TableDestination> {
+    private final String tableName;
+
+    public ClusteredDestinations(String tableName) {
+      this.tableName = tableName;
+    }
+
+    @Override
+    public @Nullable Coder<TableDestination> getDestinationCoder() {
+      return TableDestinationCoderV3.of();
+    }
+
+    @Override
+    public TableDestination getDestination(ValueInSingleWindow<TableRow> element) {
+      return new TableDestination(
+          String.format("%s.%s", DATASET_NAME, tableName), null, null, CLUSTERING);
+    }
+
+    @Override
+    public TableDestination getTable(TableDestination destination) {
+      return destination;
+    }
+
+    @Override
+    public TableSchema getSchema(TableDestination destination) {
+      return SCHEMA;
+    }
+  }
+
+  @Test
+  public void testE2EBigQueryClusteringNoPartitionTableFunction() throws Exception {
+    String tableName = "weather_stations_clustered_table_function_" + System.currentTimeMillis();
+
+    Pipeline p = Pipeline.create(options);
+
+    p.apply(BigQueryIO.readTableRows().from(options.getBqcInput()))
+        .apply(ParDo.of(new KeepStationNumberAndConvertDate()))
+        .apply(
+            BigQueryIO.writeTableRows()
+                .to(
+                    (ValueInSingleWindow<TableRow> vsw) ->
+                        new TableDestination(
+                            String.format("%s.%s", DATASET_NAME, tableName),
+                            null,
+                            null,
+                            CLUSTERING))
+                .withClustering(CLUSTERING)
+                .withSchema(SCHEMA)
+                .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_IF_NEEDED)
+                .withWriteDisposition(BigQueryIO.Write.WriteDisposition.WRITE_TRUNCATE));
+
+    p.run().waitUntilFinish();
+
+    Table table = bqClient.tables().get(options.getProject(), DATASET_NAME, tableName).execute();
+
+    Assert.assertEquals(CLUSTERING, table.getClustering());
```

Comment:
Can we also verify the data written ?

Reply:
Sure ...  I had set a default table expiration of 5 days, to not unnecessarily store IT data.  Currently that is 5 days -- do we think we should have a longer retention period?

Reply:
The data is being written/created (verified manually).    Do you have a suggestion for an Assert to use ( check row count,?check a specific value? ).  Clustering *should* only effect the table being created and not the data itself, but doesn't hurt to add some additional checks/tests/assertions.  

Perhaps there is an example from another IT that you think would be good?

Reply:
Yeah, I mean to check the properties of the data (for example, rowcount, a checksum of the data) in the same IT using an assert. There's an example here: https://github.com/apache/beam/blob/419c03c71d47bc1a4001735e260fcfe083da164d/examples/java/src/test/java/org/apache/beam/examples/cookbook/BigQueryTornadoesIT.java#L70

Reply:
ya, that makes sense to have.  Rowcount super straightforward.  Hash is better, but ... I don't see a great way to easily generate to put into the test ( get it the first time ).  I'll think more on the approach today.  Thanks!

Reply:
I implimented # of lines and expected bytes -- This seemed sufficient and was easier to implement than using bigquerymatcher.